vm/qemu/snapshot_linux.go

```diff
@@ -0,0 +1,246 @@
+	// ivshmem-doorbell expects an external server that communicates via a unix socket.
+	// The protocol is not documented, for details see:
+	// https://github.com/qemu/qemu/blob/master/hw/misc/ivshmem.c
+	// https://github.com/qemu/qemu/blob/master/contrib/ivshmem-server/ivshmem-server.c
+	conn, err := inst.ivsListener.AcceptUnix()
+	if err != nil {
+		return fmt.Errorf("qemu: unix accept failed: %w", err)
+	}
+	inst.ivsListener.Close()
+	inst.ivsListener = nil
+	inst.ivsConn = conn
+
+	msg := make([]byte, 8)
+	// Send protocol version 0.
+	binary.LittleEndian.PutUint64(msg, 0)
+	if _, err := conn.Write(msg); err != nil {
+		return fmt.Errorf("qemu: ivs conn write failed: %w", err)
+	}
+	// Send VM id 0.
+	binary.LittleEndian.PutUint64(msg, 0)
+	if _, err := conn.Write(msg); err != nil {
+		return fmt.Errorf("qemu: ivs conn write failed: %w", err)
+	}
+	// Send shared memory file FD.
+	binary.LittleEndian.PutUint64(msg, ^uint64(0))
+	rights := syscall.UnixRights(inst.doorbellFD)
+	if _, _, err := conn.WriteMsgUnix(msg, rights, nil); err != nil {
+		return fmt.Errorf("qemu: ivs conn sendmsg failed: %w", err)
+	}
+	// Send event FD for VM 1 interrupt vector 0.
+	binary.LittleEndian.PutUint64(msg, 1)
+	rights = syscall.UnixRights(inst.eventFD)
+	if _, _, err := conn.WriteMsgUnix(msg, rights, nil); err != nil {
+		return fmt.Errorf("qemu: ivs conn sendmsg failed: %w", err)
+	}
+	return nil
+}
+
+func (inst *instance) SetupSnapshot(input []byte) error {
+	copy(inst.input, input)
+	// Tell executor that we are ready to snapshot and wait for an ack.
+	inst.header.UpdateState(flatrpc.SnapshotStateHandshake)
+	if !inst.waitSnapshotStateChange(flatrpc.SnapshotStateHandshake, 10*time.Minute) {
+		return fmt.Errorf("executor does not start snapshot handshake")
+	}
+	if _, err := inst.hmp("migrate_set_capability x-ignore-shared on", 0); err != nil {
+		return err
+	}
+	if _, err := inst.hmp("savevm syz", 0); err != nil {
+		return err
+	}
+	if inst.debug {
+		inst.hmp("info snapshots", 0) // this prints size of the snapshot
+	}
+	inst.header.UpdateState(flatrpc.SnapshotStateSnapshotted)
+	if !inst.waitSnapshotStateChange(flatrpc.SnapshotStateSnapshotted, time.Minute) {
+		return fmt.Errorf("executor has not confirmed snapshot handshake")
+	}
+	return nil
+}
+
+func (inst *instance) RunSnapshot(timeout time.Duration, input []byte) (result, output []byte, err error) {
+	copy(inst.input, input)
+	inst.header.OutputOffset = 0
+	inst.header.OutputSize = 0
+	inst.header.UpdateState(flatrpc.SnapshotStateExecute)
+	if _, err := inst.hmp("loadvm syz", 0); err != nil {
+		return nil, nil, err
+	}
+	inst.waitSnapshotStateChange(flatrpc.SnapshotStateExecute, timeout)
+	resStart := int(flatrpc.ConstMaxInputSize) + int(atomic.LoadUint32(&inst.header.OutputOffset))
+	resEnd := resStart + int(atomic.LoadUint32(&inst.header.OutputSize))
+	var res []byte
+	if resEnd <= len(inst.shmem) {
+		res = inst.shmem[resStart:resEnd:resEnd]
```

Comment:
Should we also check for `resStart < resEnd`? May this panic in case of memory corruption otherwise?

Reply:
The values we load OutputOffset/OutputSize are 32-bit and unsigned, so they cannot overflow 64-bit int.